src/main/scala/sigmastate/eval/Evaluation.scala

```diff
@@ -84,58 +89,117 @@ trait Evaluation extends RuntimeCosting { IR: IRContext =>
-         ContextM.getVar(_,_,_)  => true
-    case SigmaM.propBytes(_) => true
-    case _: CReplCollCtor[_] | _: PairOfColsCtor[_,_] => true
-    case CollM.length(_) | CollM.map(_,_) | CollM.sum(_,_) | CollM.zip(_,_) | CollM.slice(_,_,_) | CollM.apply(_,_) |
-         CollM.append(_,_) | CollM.foldLeft(_,_,_) => true
-    case CBM.replicate(_,_,_) | CBM.fromItems(_,_,_) => true
-    case BoxM.propositionBytes(_) | BoxM.bytesWithoutRef(_) | BoxM.getReg(_,_,_) => true
-    case OM.get(_) | OM.getOrElse(_,_) | OM.fold(_,_,_) | OM.isDefined(_) => true
-    case _: CostOf | _: SizeOf[_] => true
-    case _: Upcast[_,_] => true
-    case _: Apply[_,_] => true
-    case SPCM.some(_) => true
-    case _ => false
+  private val _allowedOpCodesInCosting = HashSet[OpCode](
+    Undefined,
+    AppendCode,
+    ByIndexCode,
+    ConstantCode,
+    DivisionCode,
+    DowncastCode,
+    ExtractBytesWithNoRefCode,
+    ExtractRegisterAs,
+    ExtractScriptBytesCode,
+    FoldCode,
+    FuncApplyCode,
+    FuncValueCode,
+    GetVarCode,
+    InputsCode,
+    LastBlockUtxoRootHashCode,
+    MapCollectionCode,
+    MaxCode,
+    MethodCallCode,
+    MinCode,
+    MinusCode,
+    ModuloCode,
+    MultiplyCode,
+    OptionGetCode,
+    OptionGetOrElseCode,
+    OptionIsDefinedCode,
+    OutputsCode,
+    PlusCode,
+    SelectFieldCode,
+    SelfCode,
+    SigmaPropBytesCode,
+    SizeOfCode,
+    SliceCode,
+    TupleCode,
+    UpcastCode,
+  )
+
+  def isAllowedOpCodeInCosting(opCode: OpCode): Boolean = _allowedOpCodesInCosting.contains(opCode)
+
+  object isNonSerializable {
+    def unapply(op: Def[_]): Option[Unit] = op match {
+      case _: OpCost | _: PerKbCostOf | _: Cast[_] => Some()
+      case _: IntPlusMonoid => Some()
+      case _: ThunkDef[_] => Some()
+      case SCM.inputs(_) | SCM.outputs(_) | SCM.dataInputs(_) | SCM.selfBox(_) | SCM.lastBlockUtxoRootHash(_) |
+           SCM.headers(_) | SCM.preHeader(_) | SCM.getVar(_, _, _) => Some()
+      case SBM.propositionBytes(_) | SBM.bytes(_) | SBM.bytesWithoutRef(_) | SBM.registers(_) |
+           SBM.getReg(_, _, _) | SBM.tokens(_) => Some()
+      case SSPM.propBytes(_) => Some()
+      case SAVM.tVal(_) | SAVM.valueSize(_) => Some()
+      case SizeM.dataSize(_) => Some()
+      case SPairM.l(_) | SPairM.r(_) => Some()
+      case SCollM.sizes(_) => Some()
+      case SOptM.sizeOpt(_) => Some()
+      case SFuncM.sizeEnv(_) => Some()
+      case _: CSizePairCtor[_, _] | _: CSizeFuncCtor[_, _, _] | _: CSizeOptionCtor[_] | _: CSizeCollCtor[_] |
+           _: CSizeBoxCtor | _: CSizeContextCtor | _: CSizeAnyValueCtor => Some()
+      case _: CReplCollCtor[_] | _: PairOfColsCtor[_, _] => Some()
+      case CollM.sum(_, _) => Some()
+      case CBM.replicate(_, _, _) | CBM.fromItems(_, _, _) => Some()
+      case _: CostOf | _: SizeOf[_] => Some()
+      case SPCM.some(_) => Some()
+      case _ => None
+    }
```

Comment:
@greenhat the same performance argument applies here. 
Please avoid allocations.
Also here another possible optimization. 
Instead of checking each individual method we can check it's class
this will reduce the number of cases in `match`, hence reduce the number of the checks.

Reply:
Done.